### PR TITLE
fix: refactor navigation on User Profile Screen and fix avatar upload bugs [AR-1204][AR-1210][AR-1211]

### PR DESCRIPTION
**Pull request by gongracr**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/AR-1204" title="AR-1204" target="_blank">SubtaskAR-1204</a>  Refactor navigation between UserProfile and Avatar Picker screens
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

- Currently the navigation between the `UserProfileScreen` and the `AvatarPickerScreen` was made via the `UserProfileViewModel` instead of using its own View Model.
- The way we had previously to share the picked image on the gallery/camera was also broken, hence not allowing to update the Avatar Icon on the `UserProfileScreen`. Also, the Camera option, was taking pictures on preview mode, hence pixelating the image too much.
- The User Profile didn't have any persistence load/save mechanisms, making the whole process of picking a new image, only available during the ongoing session.

### Solutions

- Added a new `AvatarPickerViewModel` to deal with the avatar picker logic.
- Added a new file provider to allow the Android Camera save the taken picture on a specific internal path.
- Delegated the process of uploading the picture to the `AvatarPickerViewModel`.
- Thanks to this [PR](https://github.com/wireapp/kalium/pull/243) I could refactored how the avatar image is retrieved and loaded (via the `GetPublicAssetUseCase`) and persisted to memory once a new picture has been chosen/taken (via the `UploadUserAvatarUseCase`) on the corresponding View Models.
- Added [Coil library](https://coil-kt.github.io/coil/) to handle image loading on the app.

### Attachments (Optional)
https://user-images.githubusercontent.com/2468164/157436359-c72e4228-969a-489c-a6ec-b7d58932d3ea.mov

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [x] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [x] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: app/src/main/kotlin/com/wire/android/ui/userprofile/image/AvatarPickerViewModel.kt
Comment (yamilmedina) [inline app/src/main/kotlin/com/wire/android/ui/userprofile/image/AvatarPickerViewModel.kt:55]: I think for now it's ok doing it fixed to "image/jpg", but later on (for other types of assets) we could manage to extract the actual mimetype from the Uri, maybe as an Extension Function on `FileUtil` class. 
Using: 
- https://developer.android.com/reference/android/webkit/MimeTypeMap#getFileExtensionFromUrl(java.lang.String)
- https://developer.android.com/reference/android/webkit/MimeTypeMap#getMimeTypeFromExtension(java.lang.String)

**Review by yamilmedina — DISMISSED** [review]
Awesome work! 👏  a pain in the a$$ working the FileProvider, but now it's working.

Just a small task before merging, taking care of this detekt issues, just a suppress will do it:
https://github.com/wireapp/wire-android-reloaded/runs/5482156311?check_suite_focus=true 

exceptions - 40min debt
	TooGenericExceptionCaught - [e] at /home/runner/work/wire-android-reloaded/wire-android-reloaded/app/src/main/kotlin/com/wire/android/ui/userprofile/UserProfileViewModel.kt:95:18
	TooGenericExceptionCaught - [e] at /home/runner/work/wire-android-reloaded/wire-android-reloaded/app/src/main/kotlin/com/wire/android/ui/userprofile/image/AvatarPickerViewModel.kt:49:18
style - 10min debt
	MagicNumber - [toByteArray] at /home/runner/work/wire-android-reloaded/wire-android-reloaded/app/src/main/kotlin/com/wire/android/util/FileUtil.kt:33:84

File: app/src/main/kotlin/com/wire/android/navigation/NavigationItem.kt
Comment (saleniuk) [inline app/src/main/kotlin/com/wire/android/navigation/NavigationItem.kt:48]: I think it can be removed because it isn't used anywhere now as we declare arguments directly in `getRouteWithArgs` and we don't have any non-string arguments - what do you think?

File: app/src/main/kotlin/com/wire/android/ui/userprofile/UserProfileViewModel.kt
Comment (gongracr) [reply app/src/main/kotlin/com/wire/android/ui/userprofile/UserProfileViewModel.kt:56]: The idea was simply to cache the asset id on the data store and that way save one more call to the `getSelf` api and only make a call to the `GetPublicAssetUseCase` which will return it from memory in case it had been also cached previously

Comment (saleniuk) [reply app/src/main/kotlin/com/wire/android/ui/userprofile/UserProfileViewModel.kt:56]: I just feel like maybe we should rather cache it on the repository level, it would be easier to use - what do you think?

Comment (gongracr) [reply app/src/main/kotlin/com/wire/android/ui/userprofile/UserProfileViewModel.kt:56]: But this is already happening with `downloadPublicAsset` method, isn't it? But we still need to keep somewhere a reference/id to the specific asset we cached previously and that we want to retrieve again

Comment (gongracr) [reply app/src/main/kotlin/com/wire/android/ui/userprofile/UserProfileViewModel.kt:56]: Or you meant, also adding methods to save/load the avatar asset id, or directly the self user avatar?

File: app/src/main/kotlin/com/wire/android/ui/userprofile/UserProfileViewModel.kt
Comment (gongracr) [reply app/src/main/kotlin/com/wire/android/ui/userprofile/UserProfileViewModel.kt:178]: For now I'd prefer to keep it temporarily in this file until we get a common view on how to handle core/network failures in a unified way.  Also, moving it to `integers.xml` would require to pass a Context instance to the viewmodel method, which I'm not really fond of. Would that still be ok?

File: app/src/main/kotlin/com/wire/android/ui/common/UserProfileAvatar.kt
Comment (trOnk12) [inline app/src/main/kotlin/com/wire/android/ui/common/UserProfileAvatar.kt:56]: could you at least change it to Color.Black :P you do not need calling Companion there

File: app/src/main/kotlin/com/wire/android/ui/home/HomeTopBar.kt
Comment (trOnk12) [inline app/src/main/kotlin/com/wire/android/ui/home/HomeTopBar.kt:45]: hmmm... what did you try to achieve with it using viewModel as a key ?

File: app/src/main/kotlin/com/wire/android/ui/home/HomeTopBar.kt
Comment (trOnk12) [inline app/src/main/kotlin/com/wire/android/ui/home/HomeTopBar.kt:49]: I changed a lot this HomeTopBar function in my PR so I guess I am going to refactor it anyway, but I think it would make more sense to just pass bytearray here and hoist loading it up to the parent

File: app/src/main/kotlin/com/wire/android/ui/home/HomeViewModel.kt
Comment (trOnk12) [inline app/src/main/kotlin/com/wire/android/ui/home/HomeViewModel.kt:57]: are we handeling here errors ? you just cast it to  PublicAssetResult.Success, which is abit dangerous I think here ;p

File: app/src/main/kotlin/com/wire/android/ui/userprofile/UserProfileViewModel.kt
Comment (trOnk12) [inline app/src/main/kotlin/com/wire/android/ui/userprofile/UserProfileViewModel.kt:53]: we are not handeling any failures here

File: app/src/main/kotlin/com/wire/android/ui/userprofile/UserProfileViewModel.kt
Comment (trOnk12) [inline app/src/main/kotlin/com/wire/android/ui/userprofile/UserProfileViewModel.kt:73]: In my opinion a better user experience would be if we load the avatar seperetly as it may take more time because we are calling additional usecase for that, we could display user info immediatly though, what do you think ?

File: app/src/main/kotlin/com/wire/android/ui/userprofile/image/AvatarPicker.kt
Comment (trOnk12) [inline app/src/main/kotlin/com/wire/android/ui/userprofile/image/AvatarPicker.kt:56]: I think we talked about this on the call, maybe I did not explain it correctly to you but here on the designs : 

image

the loading of the image is happening actually on the UserProfileScreen. I also discussed this approach with Vitor I think you took and the issues here are : 

if loading takes long time, it looks like screen freezes, we could add a spinner on AvatarPicker to indicate loading, but then lets say either user has no internet or it takes like 1 minutes, we have no option to go back to the previous screeen while loading as the back button is actually starting the upload.

**Review by trOnk12 — COMMENTED** [review]
Generally, good job 🥇left some comments and questions 

also the picture on the profile screen looks really weird : 

image

File: app/src/main/kotlin/com/wire/android/ui/home/HomeViewModel.kt
Comment (gongracr) [reply app/src/main/kotlin/com/wire/android/ui/home/HomeViewModel.kt:57]: Good catch. Adding a dummy try/catch clause for now. In any case this smart casts will be soon refactored once we decide on a better wrapping for these use cases results

File: app/src/main/kotlin/com/wire/android/ui/userprofile/image/AvatarPicker.kt
Comment (gongracr) [reply app/src/main/kotlin/com/wire/android/ui/userprofile/image/AvatarPicker.kt:56]: I know that the designs indicate that we should do the upload on the user profile screen, but actually from the technical point of view, it would make more sense to do the api call to upload the avatar on the picker screen. 

In any case, my idea was to just merge these changes now and fix the picker screen as it is not functional atm. Then once Wolfgang is back, we will need to have a talk with him, and decide on what's the best approach on where to upload it, and error handling, since the current design also doesn't allow the user to reset the picture to the original one once she/he has picked a new one. What do you think?

File: app/src/main/kotlin/com/wire/android/ui/userprofile/UserProfileViewModel.kt
Comment (gongracr) [reply app/src/main/kotlin/com/wire/android/ui/userprofile/UserProfileViewModel.kt:53]: If the avatar asset raw is null, then the default fallback picture will be shown

Comment (trOnk12) [reply app/src/main/kotlin/com/wire/android/ui/userprofile/UserProfileViewModel.kt:53]: that is true, but we are calling : getPublicAssetUseCase(avatarAssetId) inside getAvatarAsByteArrayOrNull() so there could some failure from the use case and we are not handeling but we are just displaying the default picture and not show any snackbar or anything, is it something we want to do ?

File: app/src/main/kotlin/com/wire/android/ui/userprofile/image/AvatarPicker.kt
Comment (trOnk12) [reply app/src/main/kotlin/com/wire/android/ui/userprofile/image/AvatarPicker.kt:56]: sure, I am fine with it, just was thinking you did notice it

File: app/src/main/kotlin/com/wire/android/ui/userprofile/UserProfileViewModel.kt
Comment (gongracr) [reply app/src/main/kotlin/com/wire/android/ui/userprofile/UserProfileViewModel.kt:73]: That is indeed a good point. However I think 99% of the cases it will take more or less the same time to load both, as avatar is meant to change rarely and therefore the subsequent calls to this use case will be rather immediate as the avatar will be pulled directly from the assets database. I could still change the logic a bit, to load the asset on a separate coroutine, if you think it's indeed necessary and the current avatar loading times are too high though.

File: app/src/main/kotlin/com/wire/android/ui/userprofile/UserProfileViewModel.kt
Comment (gongracr) [reply app/src/main/kotlin/com/wire/android/ui/userprofile/UserProfileViewModel.kt:53]: The error here will be when downloading the avatar picture, not when trying to upload it as the design specs suggest, so not sure what should I show here, maybe some snack bar saying "There was an error while fetching the user info, check your internet connection" ? But it is also true that I didn't add the snackbar when the uploads fails so will do it right away.

File: app/src/main/kotlin/com/wire/android/ui/userprofile/UserProfileViewModel.kt
Comment (saleniuk) [reply app/src/main/kotlin/com/wire/android/ui/userprofile/UserProfileViewModel.kt:56]: Yeah, can't we also cache this asset id in the repository? 😄

File: app/src/main/kotlin/com/wire/android/ui/userprofile/UserProfileViewModel.kt
Comment (trOnk12) [reply app/src/main/kotlin/com/wire/android/ui/userprofile/UserProfileViewModel.kt:53]: oki doki, thanks for explanation, your call on that one :P I am fine with both

File: app/src/main/kotlin/com/wire/android/ui/userprofile/UserProfileViewModel.kt
Comment (trOnk12) [reply app/src/main/kotlin/com/wire/android/ui/userprofile/UserProfileViewModel.kt:73]: I do not think it can hurt and may improve a user experience abit, but your call if you think it is worth it and will not take much time

File: app/src/main/kotlin/com/wire/android/ui/userprofile/UserProfileViewModel.kt
Comment (gongracr) [reply app/src/main/kotlin/com/wire/android/ui/userprofile/UserProfileViewModel.kt:56]: In any case I will mark it with a TODO: to adapt to this once we revisit this for the File sharing feature, where we will most likely create another db table to store the assets ids

File: app/src/main/kotlin/com/wire/android/ui/home/HomeTopBar.kt
Comment (gongracr) [reply app/src/main/kotlin/com/wire/android/ui/home/HomeTopBar.kt:45]: Added some inline comment to clarify it

File: app/src/main/kotlin/com/wire/android/ui/userprofile/UserProfileViewModel.kt
Comment (gongracr) [reply app/src/main/kotlin/com/wire/android/ui/userprofile/UserProfileViewModel.kt:73]: Good catch. Added

File: app/src/main/kotlin/com/wire/android/ui/userprofile/UserProfileViewModel.kt
Comment (trOnk12) [inline app/src/main/kotlin/com/wire/android/ui/userprofile/UserProfileViewModel.kt:96]: I just now remember that while I was working on it the use cases do not uses the DIspatcher.IO so I would suggest to use it here instead so that we do not freeze the UI

File: app/src/main/kotlin/com/wire/android/ui/userprofile/UserProfileViewModel.kt
Comment (trOnk12) [inline app/src/main/kotlin/com/wire/android/ui/userprofile/UserProfileViewModel.kt:81]: I woudl suggest to add here withContext(Dispatchers.IO)

File: app/src/main/kotlin/com/wire/android/ui/userprofile/UserProfileViewModel.kt
Comment (trOnk12) [inline app/src/main/kotlin/com/wire/android/ui/userprofile/UserProfileViewModel.kt:182]: how about passing Int value that do not have much meaning, we pass a sealed class or an enum ?

**Review by trOnk12 — APPROVED** [review]
small comments but looks good to me ! goob job 👍 🥇

**Review by trOnk12 — APPROVED** [review]
well done 🍰

**Review by trOnk12 — APPROVED** [review]
even better 💪

**Pull request merged**
